src/utransport.rs

```diff
@@ -0,0 +1,647 @@
+                let topic = attributes.clone().sink;
+                let zenoh_key = UPClientZenoh::to_zenoh_key_string(&topic)?;
+                // Send Publish
+                self.send_publish(&zenoh_key, payload, attributes).await
+            }
+            UMessageType::UMESSAGE_TYPE_REQUEST => {
+                Validators::Request
+                    .validator()
+                    .validate(&attributes)
+                    .map_err(|e| {
+                        UStatus::fail_with_code(
+                            UCode::INVALID_ARGUMENT,
+                            format!("Wrong Request UAttributes {e:?}"),
+                        )
+                    })?;
+                // Get Zenoh key
+                let topic = attributes.clone().sink;
+                let zenoh_key = UPClientZenoh::to_zenoh_key_string(&topic)?;
+                // Send Request
+                self.send_request(&zenoh_key, payload, attributes).await
+            }
+            UMessageType::UMESSAGE_TYPE_RESPONSE => {
+                Validators::Response
+                    .validator()
+                    .validate(&attributes)
+                    .map_err(|e| {
+                        UStatus::fail_with_code(
+                            UCode::INVALID_ARGUMENT,
+                            format!("Wrong Response UAttributes {e:?}"),
+                        )
+                    })?;
+                // Get Zenoh key
+                let topic = attributes.clone().source;
+                let zenoh_key = UPClientZenoh::to_zenoh_key_string(&topic)?;
+                // Send Response
+                self.send_response(&zenoh_key, payload, attributes).await
+            }
+            UMessageType::UMESSAGE_TYPE_UNSPECIFIED => Err(UStatus::fail_with_code(
+                UCode::INVALID_ARGUMENT,
+                "Wrong Message type in UAttributes",
+            )),
+        }
+    }
+
+    async fn receive(&self, _topic: UUri) -> Result<UMessage, UStatus> {
+        Err(UStatus::fail_with_code(
+            UCode::UNIMPLEMENTED,
+            "Not implemented",
+        ))
+    }
+
+    async fn register_listener(
+        &self,
+        topic: UUri,
+        listener: Box<dyn Fn(Result<UMessage, UStatus>) + Send + Sync + 'static>,
+    ) -> Result<String, UStatus> {
+        let listener = Arc::new(listener);
+        if topic.authority.is_some() && topic.entity.is_none() && topic.resource.is_none() {
+            // This is special UUri which means we need to register both listeners
+            // RPC response
+            let mut listener_str = format!(
+                "{}_{:X}",
+                String::from("up/") + &UPClientZenoh::get_uauth_from_uuri(&topic)? + "/**",
+                self.callback_counter.fetch_add(1, Ordering::SeqCst)
+            );
+            self.rpc_callback_map
+                .lock()
+                .unwrap()
+                .insert(listener_str.clone(), listener.clone());
```

Comment:
Yeah, this makes sense. I think this should work for the uStreamer use-case. :+1: 

I see you handled the `unregister_listener()` nicely too, so we can provide the concatenated string back as-is and you'll break it and unregister each part. Nice.